File: bci/api/typescript/src/transport/WebSocketTransport.ts
/**
 * WIA BCI WebSocket Transport
 *
 * WebSocket-based transport implementation.
 */

import { BaseTransport } from './ITransport';
import { BciMessage, TransportOptions, HeartbeatConfig, ReconnectConfig } from '../protocol/types';
import { MessageParser, serialize, serializeBinary } from '../protocol/MessageParser';

/**
 * WebSocket transport options
 */
export interface WebSocketTransportOptions extends TransportOptions {
  protocols?: string[];
  binaryMode?: boolean;
  heartbeat?: HeartbeatConfig;
  reconnect?: ReconnectConfig;
}

/**
 * Default configurations
 */
const DEFAULT_HEARTBEAT: HeartbeatConfig = {
  pingInterval: 30000,
  pongTimeout: 10000,
  maxMissedPongs: 3,
};

const DEFAULT_RECONNECT: ReconnectConfig = {
  enabled: true,
  maxAttempts: 5,
  initialDelay: 1000,
  maxDelay: 30000,
  backoffMultiplier: 2,
};

/**
 * WebSocket transport implementation
 */
export class WebSocketTransport extends BaseTransport {
  private ws: WebSocket | null = null;
  private parser: MessageParser;
  private url: string = '';
  private binaryMode: boolean = false;
  private heartbeatConfig: HeartbeatConfig;
  private reconnectConfig: ReconnectConfig;
  private pingTimer: ReturnType<typeof setInterval> | null = null;
  private pongTimer: ReturnType<typeof setTimeout> | null = null;
  private reconnectAttempts: number = 0;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private missedPongs: number = 0;
  private intentionalClose: boolean = false;

  constructor(options: WebSocketTransportOptions = {}) {
    super();
    this.parser = new MessageParser();
    this.binaryMode = options.binaryMode ?? false;
    this.heartbeatConfig = { ...DEFAULT_HEARTBEAT, ...options.heartbeat };
    this.reconnectConfig = { ...DEFAULT_RECONNECT, ...options.reconnect };
  }

  /**
   * Connect to WebSocket server
   */
  async connect(url: string, options?: WebSocketTransportOptions): Promise<void> {
    if (this.state !== 'disconnected') {
      throw new Error('Transport already connected or connecting');
    }

    this.url = url;
    this.intentionalClose = false;

    if (options?.binaryMode !== undefined) {
      this.binaryMode = options.binaryMode;
    }
    if (options?.heartbeat) {
      this.heartbeatConfig = { ...this.heartbeatConfig, ...options.heartbeat };
    }
    if (options?.reconnect) {
      this.reconnectConfig = { ...this.reconnectConfig, ...options.reconnect };
    }

    return this.doConnect();
  }

  /**
   * Internal connect
   */
  private doConnect(): Promise<void> {
    return new Promise((resolve, reject) => {
      this.state = 'connecting';

      try {
        this.ws = new WebSocket(this.url, ['wia-bci-v1']);
        this.ws.binaryType = 'arraybuffer';

        this.ws.onopen = () => {
          this.state = 'connected';
          this.reconnectAttempts = 0;
          this.startHeartbeat();
          this.emitOpen();
          resolve();
        };

        this.ws.onclose = (event) => {
          this.handleClose(event.code, event.reason);
        };

        this.ws.onerror = (event) => {
          const error = new Error('WebSocket error');
          this.emitError(error);
          if (this.state === 'connecting') {
            reject(error);
          }
        };

        this.ws.onmessage = (event) => {
          this.handleMessage(event.data);
        };
      } catch (err) {
        this.state = 'disconnected';
        reject(err);
      }
    });
  }

  /**
   * Disconnect from server
   */
  async disconnect(): Promise<void> {
    this.intentionalClose = true;
    this.stopHeartbeat();
    this.clearReconnectTimer();

    if (this.ws && this.state === 'connected') {
      this.state = 'closing';
      this.ws.close(1000, 'Client disconnect');
    }

    this.state = 'disconnected';
    this.ws = null;
  }

  /**
   * Send message
   */
  async send(message: BciMessage): Promise<void> {
    if (!this.ws || this.state !== 'connected') {
      throw new Error('Not connected');
    }

    if (this.binaryMode) {
      const buffer = serializeBinary(message);
      this.ws.send(buffer);
    } else {
      const json = serialize(message);
      this.ws.send(json);
    }
  }

  /**
   * Handle incoming message
   */
  private handleMessage(data: string | ArrayBuffer): void {
    const result = this.parser.parse(data);

    if (result.success && result.message) {
      // Handle pong internally
      if (result.message.type === 'pong') {
        this.handlePong();
        return;
      }

      this.emitMessage(result.message);
    } else if (result.error) {
      this.emitError(new Error(result.error.message));
    }
  }

  /**
   * Handle connection close
   */
  private handleClose(code: number, reason: string): void {
    this.stopHeartbeat();

    if (this.intentionalClose) {
      this.state = 'disconnected';
      this.emitClose(code, reason);
      return;
    }

    // Attempt reconnection
    if (this.reconnectConfig.enabled && this.reconnectAttempts < this.reconnectConfig.maxAttempts) {
      this.attemptReconnect();
    } else {
      this.state = 'disconnected';
      this.emitClose(code, reason);
    }
  }

  /**
   * Attempt reconnection
   */
  private attemptReconnect(): void {
    this.reconnectAttempts++;

    const delay = Math.min(
      this.reconnectConfig.initialDelay *
        Math.pow(this.reconnectConfig.backoffMultiplier, this.reconnectAttempts - 1),
      this.reconnectConfig.maxDelay
    );

    this.reconnectTimer = setTimeout(async () => {
      try {
        await this.doConnect();
      } catch {
        // Will be handled by handleClose
      }
    }, delay);
  }

  /**
   * Clear reconnect timer
   */
  private clearReconnectTimer(): void {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
  }

  /**
   * Start heartbeat
   */
  private startHeartbeat(): void {
    this.missedPongs = 0;

    this.pingTimer = setInterval(() => {
      this.sendPing();
    }, this.heartbeatConfig.pingInterval);
  }

  /**
   * Stop heartbeat
   */
  private stopHeartbeat(): void {
    if (this.pingTimer) {
      clearInterval(this.pingTimer);
      this.pingTimer = null;
    }
    if (this.pongTimer) {
      clearTimeout(this.pongTimer);
      this.pongTimer = null;
    }
  }

  /**
   * Send ping
   */
  private sendPing(): void {
    if (!this.ws || this.state !== 'connected') return;

    const ping = {
      protocol: 'wia-bci' as const,
      version: '1.0.0',
      messageId: `ping-${Date.now()}`,
      timestamp: Date.now(),
      type: 'ping' as const,
      payload: { clientTime: Date.now() },
    };

    try {
      this.ws.send(serialize(ping));

      // Start pong timeout
      this.pongTimer = setTimeout(() => {
        this.handlePongTimeout();
      }, this.heartbeatConfig.pongTimeout);
    } catch {
      // Connection may be lost
    }
  }

  /**
   * Handle pong response
   */
  private handlePong(): void {
    this.missedPongs = 0;
    if (this.pongTimer) {
      clearTimeout(this.pongTimer);
      this.pongTimer = null;
    }
  }

  /**
   * Handle pong timeout
   */
  private handlePongTimeout(): void {
    this.missedPongs++;

    if (this.missedPongs >= this.heartbeatConfig.maxMissedPongs) {
      // Connection lost
      this.ws?.close(4000, 'Heartbeat timeout');
    }
  }

  /**
   * Dispose transport
   */
  dispose(): void {
    this.disconnect();
    super.dispose();
  }
}

File: bci/api/python/wia_bci/transport/websocket_transport.py
"""
WIA BCI WebSocket Transport

WebSocket-based transport implementation.
"""

import asyncio
from dataclasses import dataclass, field
from typing import Optional

from .base_transport import BaseTransport, TransportState
from ..protocol.types import (
    BciMessage,
    TransportOptions,
    HeartbeatConfig,
    ReconnectConfig,
)
from ..protocol.message_parser import MessageParser, serialize, serialize_binary
from ..protocol.message_builder import MessageBuilder

# Optional websockets import
try:
    import websockets
    from websockets.client import WebSocketClientProtocol

    WEBSOCKETS_AVAILABLE = True
except ImportError:
    WEBSOCKETS_AVAILABLE = False
    WebSocketClientProtocol = None


@dataclass
class WebSocketTransportOptions(TransportOptions):
    """WebSocket transport options."""

    protocols: list = field(default_factory=lambda: ["wia-bci-v1"])
    binary_mode: bool = False
    heartbeat: Optional[HeartbeatConfig] = None
    reconnect: Optional[ReconnectConfig] = None


# Default configurations
DEFAULT_HEARTBEAT = HeartbeatConfig(
    ping_interval=30000,
    pong_timeout=10000,
    max_missed_pongs=3,
)

DEFAULT_RECONNECT = ReconnectConfig(
    enabled=True,
    max_attempts=5,
    initial_delay=1000,
    max_delay=30000,
    backoff_multiplier=2.0,
)


class WebSocketTransport(BaseTransport):
    """WebSocket transport implementation."""

    def __init__(self, options: Optional[WebSocketTransportOptions] = None):
        super().__init__()
        if not WEBSOCKETS_AVAILABLE:
            raise ImportError(
                "websockets package is required for WebSocketTransport. "
                "Install with: pip install websockets"
            )

        self._ws: Optional[WebSocketClientProtocol] = None
        self._parser = MessageParser()
        self._builder = MessageBuilder()
        self._url: str = ""
        self._binary_mode: bool = options.binary_mode if options else False
        self._heartbeat_config = options.heartbeat if options and options.heartbeat else DEFAULT_HEARTBEAT
        self._reconnect_config = options.reconnect if options and options.reconnect else DEFAULT_RECONNECT
        self._ping_task: Optional[asyncio.Task] = None
        self._receive_task: Optional[asyncio.Task] = None
        self._reconnect_attempts: int = 0
        self._missed_pongs: int = 0
        self._intentional_close: bool = False
        self._pong_received: asyncio.Event = asyncio.Event()

    async def connect(
        self, url: str, options: Optional[WebSocketTransportOptions] = None
    ) -> None:
        """Connect to WebSocket server."""
        if self._state != TransportState.DISCONNECTED:
            raise ConnectionError("Transport already connected or connecting")

        self._url = url
        self._intentional_close = False

        if options:
            if options.binary_mode is not None:
                self._binary_mode = options.binary_mode
            if options.heartbeat:
                self._heartbeat_config = options.heartbeat
            if options.reconnect:
                self._reconnect_config = options.reconnect

        await self._do_connect()

    async def _do_connect(self) -> None:
        """Internal connect."""
        self._state = TransportState.CONNECTING

        try:
            self._ws = await websockets.connect(
                self._url,
                subprotocols=["wia-bci-v1"],
            )

            self._state = TransportState.CONNECTED
            self._reconnect_attempts = 0
            self._start_heartbeat()
            self._start_receive()
            self._emit_open()

        except Exception as e:
            self._state = TransportState.DISCONNECTED
            raise e

    async def disconnect(self) -> None:
        """Disconnect from server."""
        self._intentional_close = True
        self._stop_heartbeat()
        self._stop_receive()

        if self._ws and self._state == TransportState.CONNECTED:
            self._state = TransportState.CLOSING
            await self._ws.close(1000, "Client disconnect")

        self._state = TransportState.DISCONNECTED
        self._ws = None

    async def send(self, message: BciMessage) -> None:
        """Send message."""
        if not self._ws or self._state != TransportState.CONNECTED:
            raise ConnectionError("Not connected")

        if self._binary_mode:
            data = serialize_binary(message)
            await self._ws.send(data)
        else:
            data = serialize(message)
            await self._ws.send(data)

    def _start_receive(self) -> None:
        """Start receive task."""
        self._receive_task = asyncio.create_task(self._receive_loop())

    def _stop_receive(self) -> None:
        """Stop receive task."""
        if self._receive_task:
            self._receive_task.cancel()
            self._receive_task = None

    async def _receive_loop(self) -> None:
        """Receive loop."""
        try:
            async for data in self._ws:
                await self._handle_message(data)
        except websockets.ConnectionClosed as e:
            await self._handle_close(e.code, e.reason or "")
        except Exception as e:
            self._emit_error(e)

    async def _handle_message(self, data) -> None:
        """Handle incoming message."""
        result = self._parser.parse(data)

        if result.success and result.message:
            # Handle pong internally
            if result.message.type.value == "pong":
                self._handle_pong()
                return

            self._emit_message(result.message)
        elif result.error:
            self._emit_error(Exception(result.error.get("message", "Parse error")))

    async def _handle_close(self, code: int, reason: str) -> None:
        """Handle connection close."""
        self._stop_heartbeat()

        if self._intentional_close:
            self._state = TransportState.DISCONNECTED
            self._emit_close(code, reason)
            return

        # Attempt reconnection
        if (
            self._reconnect_config.enabled
            and self._reconnect_attempts < self._reconnect_config.max_attempts
        ):
            await self._attempt_reconnect()
        else:
            self._state = TransportState.DISCONNECTED
            self._emit_close(code, reason)

    async def _attempt_reconnect(self) -> None:
        """Attempt reconnection."""
        self._reconnect_attempts += 1

        delay = min(
            self._reconnect_config.initial_delay
            * (self._reconnect_config.backoff_multiplier ** (self._reconnect_attempts - 1)),
            self._reconnect_config.max_delay,
        )

        await asyncio.sleep(delay / 1000)

        try:
            await self._do_connect()
        except Exception:
            # Will be handled by _handle_close
            pass

    def _start_heartbeat(self) -> None:
        """Start heartbeat."""
        self._missed_pongs = 0
        self._ping_task = asyncio.create_task(self._heartbeat_loop())

    def _stop_heartbeat(self) -> None:
        """Stop heartbeat."""
        if self._ping_task:
            self._ping_task.cancel()
            self._ping_task = None

    async def _heartbeat_loop(self) -> None:
        """Heartbeat loop."""
        try:
            while self._state == TransportState.CONNECTED:
                await asyncio.sleep(self._heartbeat_config.ping_interval / 1000)
                await self._send_ping()
        except asyncio.CancelledError:
            pass

    async def _send_ping(self) -> None:
        """Send ping."""
        if not self._ws or self._state != TransportState.CONNECTED:
            return

        ping = self._builder.ping()

        try:
            await self.send(ping)
            self._pong_received.clear()

            # Wait for pong
            try:
                await asyncio.wait_for(
                    self._pong_received.wait(),
                    timeout=self._heartbeat_config.pong_timeout / 1000,
                )
            except asyncio.TimeoutError:
                self._handle_pong_timeout()
        except Exception:
            pass

    def _handle_pong(self) -> None:
        """Handle pong response."""
        self._missed_pongs = 0
        self._pong_received.set()

    def _handle_pong_timeout(self) -> None:
        """Handle pong timeout."""
        self._missed_pongs += 1

        if self._missed_pongs >= self._heartbeat_config.max_missed_pongs:
            # Connection lost
            if self._ws:
                asyncio.create_task(self._ws.close(4000, "Heartbeat timeout"))

    def dispose(self) -> None:
        """Dispose transport."""
        self._stop_heartbeat()
        self._stop_receive()
        super().dispose()
